### DISCUSSION TITLE
How to create NER pipeline with multiple models in Spacy 3.x version?

### DISCUSSION THREAD

**Discussion by Singhal-harsh** [explosion/spaCy#10674] [Help: Coding & Implementations] [answered]
I am loading these two models as follows:
```
nlp_en = spacy.load('en_core_web_lg')
nlp_xx = spacy.load('xx_ent_wiki_sm')
```
Now I want to use both of their ner components and have them both in a single pipeline. 
Problem is that right now using the below approach:
```
text = "Jane lives in Brussels. Santos Carlos mora no Brasil"

@Language.component("ner_xx")
def ner_xx(doc):
   doc = nlp_xx(text)
   return doc

nlp_en.add_pipe("ner_xx", after="ner")
doc = nlp_en(text)
print([(ent.text, ent.label_) for ent in doc.ents])
```
gives me output with just the ner of nlp_xx model and not a combination of both nlp_en and nlp_xx model which is what I wanted. 

Please help me correct this, and suggest the best and maybe the easiest way to do this? 

**Basically I need to use ner components of both the xx model and the en model and then package them as a single model** to use in production for my enterprise. 

Kindly help urgently.

**Comment by adrianeboyd** [comment, marked as answer]
You don't need a custom component to do this, you just need to add the `ner` component from one pipeline to another, which you can do with  [`nlp.add_pipe(source=)`](https://spacy.io/api/language#add_pipe):

```python
import spacy
nlp_en = spacy.load("en_core_web_sm")
nlp_xx = spacy.load("xx_ent_wiki_sm")
nlp_en.add_pipe("ner", name="ner_xx", source=nlp_xx)
doc = nlp_en(text)
print(doc.ents)  # (Jane, Brussels, Santos Carlos, Brasil)
```

You have to give the new component a custom name because there's already a pipeline component named `ner` in the pipeline.

Note that the order of the `ner` components matters. The second `ner` component will only add entities where the first component has not added any entities.

There's a demo project `tutorials/ner_double` that demonstrates this interaction: https://github.com/explosion/projects/tree/v3/tutorials/ner_double

**Reply by Singhal-harsh** [reply]
Hey I am getting the following erorr when adding the component as you said: 
---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

~\AppData\Local\Temp/ipykernel_2580/3338783104.py in <module>

----> 1 `nlp_en.add_pipe("ner", name="ner_xx", source=nlp_xx)`

 

~\AppData\Roaming\Python\Python38\site-packages\spacy\language.py in add_pipe(self, factory_name, name, before, after, first, last, source, config, raw_config, validate)

    754             # We're loading the component from a model. After loading the

    755             # component, we know its real factory name

    756             pipe_component, factory_name = self.create_pipe_from_source(

    757                 factory_name, source, name=name

    758             )

 

~\AppData\Roaming\Python\Python38\site-packages\spacy\language.py in create_pipe_from_source(self, source_name, source, name)

    685         # TODO: handle errors and mismatches (vectors etc.)

    686         if not isinstance(source, self.__class__):

    687             raise ValueError(Errors.E945.format(name=source_name, source=type(source)))

    688         if not source_name in source.component_names:

    689             raise KeyError(

 

**ValueError: [E945] Can't copy pipeline component 'ner' from source. Expected loaded nlp object, but got: <class 'spacy.lang.xx.MultiLanguage'>**

Only thing different is I am using the en_core_web_lg model instead of small one, do you think that is causing the above error? if not what can be the reason and how can I resolve it? 
Thanks for your help!

**Reply by Singhal-harsh** [reply]
Can the error be because the vocab is not the same for both of the models? and if yes then how can I make it same?

**Reply by adrianeboyd** [reply]
That particular error does not look it's related to the vocab or vectors. Can you show the full code you're using? (Please use code blocks with three backticks on a separate line before and after the code: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code)

Loading with `source` handles the vocab correctly, but depending which pipeline you source from/into, you might need to add an extra step related to the vectors.

This example works fine for me (with slightly different results):

```python
import spacy
nlp_en = spacy.load("en_core_web_md")
nlp_xx = spacy.load("xx_ent_wiki_sm")
nlp_en.add_pipe("ner", name="ner_xx", source=nlp_xx)
doc = nlp_en("Jane lives in Brussels. Santos Carlos mora no Brasil")
print(doc.ents)  # (Jane, Brussels, Santos Carlos mora, Brasil)
```

Can you also provide the output of `spacy info --markdown`?

**Reply by Singhal-harsh** [reply]
Hey,
So I am using almost the same code, **with a change of using large model** that is `en_core_web_lg`, rather than medium as you used.
Following is my code:
```
import spacy
nlp_en = spacy.load("en_core_web_lg")
nlp_xx = spacy.load("xx_ent_wiki_sm")
nlp_en.add_pipe("ner", name="ner_xx", source=nlp_xx)
doc = nlp_en("Jane lives in Brussels. Santos Carlos mora no Brasil")
print(doc.ents)
```
_When I run the above, I get the error as shown above._

Following is the output of spacy info --markdown:
## Info about spaCy

- **spaCy version:** 3.0.5

- **Platform:** Windows-10-*****

- **Python version:** 3.8.0

- **Pipelines:** en_core_web_lg (3.0.0), pt_core_news_md (3.0.0), xx_ent_wiki_sm (3.0.0)

Please let me know what can be the issue, and what can be done to solve it?

**Reply by adrianeboyd** [reply]
There have been a number of bug fixes and improvements since v3.0.5, so I would suggest upgrading. I tested this example with v3.2.3 on my end. It looks like the main bug fix related to sourcing from different languages is in v3.0.6, but I see that there have been a few other bug fixes since then, too.
